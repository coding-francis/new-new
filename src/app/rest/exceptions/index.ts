export * from './http-bad-error';
export * from './http-not-found-error';
export * from './http-permission-denied-error';
export * from './http-server-error';
export * from './http-unauthorized-error';
