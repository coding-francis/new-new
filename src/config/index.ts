const config = {
    appName: process.env.APP_NAME ?? 'My App',
    port: Number(process.env.APP_PORT) || 8000,
};

export class DatabaseConfig {
    public static readonly DATABASE_URL = process.env.DATABASE_URL;
    public static readonly TEST_DATABASE_URL = process.env.TEST_DATABASE_URL;
}

export default config;
