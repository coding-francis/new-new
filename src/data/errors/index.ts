export * from './utils';
export * from './invalid-input-error';
export * from './record-not-found';
export * from './internal-db-error';
export * from './must-be-unique-error';
export * from './database-basae-error';
