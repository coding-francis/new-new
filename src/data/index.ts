import { PrismaClient } from '@prisma/client/extension';
import { DataBase } from '../internal/database.ts';
import { Logger } from '../internal/logger.ts';
import UserData, { UserDataLayer } from './user-data-layer.ts';

export interface DataLayer {
    UserDataLayer: UserDataLayer;
}

/**
 * Initialize the data layer with the database and logger. This is called in the main application entrypoint eg: `bin/index.ts` or initializing in test file .
 *
 * @param {DataBase<PrismaClient>} db - The database connection object
 * @param {Logger} logger - The logger object
 *
 * @returns {DataLayer} - The data layer object
 *
 * @example
 * const db = new PrismaDatabase('dbUrl');
 * const logger = new Logger();
 * const dataLayer = initializeDataLayer(db, logger);
 */
export const initializeDataLayer = (
    db: DataBase<PrismaClient>,
    logger: Logger
): DataLayer => {
    return {
        UserDataLayer: new UserData(db, logger),
    };
};
