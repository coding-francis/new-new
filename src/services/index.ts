export * from './initializer';
export * from './user-service';
