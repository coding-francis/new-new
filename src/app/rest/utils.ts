import {
    BaseServiceLayerError,
    ResourceNotFoundServiceError,
    DataValidationError,
} from '../../services/errors';
import { HttpBadError, HttpNotFoundError, HttpServerError } from './exceptions';

export function handleErrors(err: Error): Error {
    let error = new HttpServerError('Internal server error');
    if (
        err instanceof BaseServiceLayerError &&
        err instanceof ResourceNotFoundServiceError
    ) {
        error = new HttpNotFoundError(error.message, error.data);
    } else if (err instanceof DataValidationError) {
        error = new HttpBadError(error.message, error.data);
    }

    return error;
}
