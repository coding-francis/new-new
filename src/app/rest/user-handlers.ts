import { Service } from '../../services';
import {
    RawReplyDefaultExpression,
    RawRequestDefaultExpression,
    RawServerDefault,
    RouteHandler,
    RouteOptions,
} from 'fastify';
import {
    BaseServiceLayerError,
    ResourceNotFoundServiceError,
} from '../../services/errors';
import { HttpNotFoundError, HttpServerError } from './exceptions';
import { UserResource } from '../../services/resource';

interface URLParams {
    id: string;
}

//I thought of adding handler to RouteOptions but I felt I needed the handler as  a `controller` function seperated from the metadata. It involves you passing
// the service twice and also typing the handler function twice. I think it's better to have the handler as a function and then pass it to the RouteOptions.
//Always don't export handler functions, export the RouteOptions and the RouteHandler.
function fetchUserById(
    srv: Pick<Service, 'UserService'>
): RouteHandler<{ Params: URLParams }> {
    return async (req, res) => {
        try {
            const user = await srv.UserService.getUserById(
                Number(req.params.id)
            );
            res.status(200).send(user);
        } catch (error) {
            if (
                error instanceof BaseServiceLayerError &&
                error instanceof ResourceNotFoundServiceError
            ) {
                throw new HttpNotFoundError(error.message, error.data);
            }

            throw new HttpServerError('Internal server error');
        }
    };
}

export function fetchUserByIdRouteOption(
    srv: Pick<Service, 'UserService'>
): RouteOptions<
    RawServerDefault,
    RawRequestDefaultExpression,
    RawReplyDefaultExpression,
    { Params: URLParams }
> {
    return {
        handler: fetchUserById(srv),
        method: 'GET',
        url: '/users/:id',
        schema: {
            params: {
                type: 'object',
                properties: {
                    id: { type: 'string' },
                },
            },
            response: {
                200: UserResource.schema,
                404: HttpNotFoundError.schema,
                500: HttpServerError.schema,
            },
        },
    };
}
