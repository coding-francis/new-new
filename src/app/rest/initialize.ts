import { pick } from 'lodash';
import { Service } from '../../services/index.ts';
import { fetchUserByIdRouteOption } from './user-handlers.ts';

export function initializeRestHandlers(srv: Service) {
    return [fetchUserByIdRouteOption(pick(srv, 'UserService'))];
}
