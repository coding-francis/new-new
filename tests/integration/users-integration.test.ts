import { getMockLogger } from './mocks';
import PrismaDatabase from '../../src/internal/database';
import { DatabaseConfig } from '../../src/config';
import { initializeDataLayer } from '../../src/data';
import { initializeRestHandlers } from '../../src/app/rest';
import { CreateUserPayload, initializeServices } from '../../src/services';
import FastifyServer from '../../src/internal/server';

function seedUserData(db: PrismaDatabase) {
    return {
        up: () =>
            db.interface.user.create({
                data: {
                    email: 'test@test.com',
                    name: 'Test User',
                    password: 'password',
                },
            }),

        down: () => db.interface.user.deleteMany({}),
    };
}

describe('Users endpoint', () => {
    beforeEach(async () => {
        const [logger, logs] = getMockLogger();
        const prismaInstance = new PrismaDatabase(
            DatabaseConfig.TEST_DATABASE_URL
        );
        await prismaInstance.connect();

        const dataLayer = initializeDataLayer(prismaInstance, logger);
        const servicesLayer = initializeServices(dataLayer, logger);
        const server = new FastifyServer(
            initializeRestHandlers(servicesLayer),
            logger
        );

        // Make sure the database is clean
        await seedUserData(prismaInstance).down();

        // Seed the database
        await seedUserData(prismaInstance).up();

        expect.setState({ logs, server, prismaInstance });
    });

    afterEach(async () => {
        const { prismaInstance } = expect.getState();
        await seedUserData(prismaInstance).down();
    });

    test('GET /users:id should return a user', async () => {
        const { server, logs } = expect.getState();

        const response = await server.app.inject({
            method: 'GET',
            url: '/api/users/1',
        });

        expect(response.statusCode).toBe(200);
        expect(response.json()).toEqual({
            id: 1,
            email: 'test@test.com',
            name: 'Test User',
        });

        expect(logs).toMatchSnapshot();
    });

    test('POST /users should create and return a user created', async () => {
        const { server, logs } = expect.getState();

        const payload: CreateUserPayload = {
            name: 'John Doe',
            email: 'email@test.com',
            password: 'password',
        };

        const response = await server.app.inject({
            method: 'POST',
            url: '/api/users',
            payload,
        });

        expect(response.statusCode).toBe(201);
        const userResponse = response.json();
        //We can't predict user id so let's delete it
        delete userResponse.id;
        expect(userResponse).toEqual({
            email: 'email@test.com',
            name: 'John Doe',
        });

        expect(logs).toMatchSnapshot('logs');
    });
});
