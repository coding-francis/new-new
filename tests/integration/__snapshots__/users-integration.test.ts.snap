// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Users endpoint GET /users:id should return a user 1`] = `
[
  {
    "level": 30,
    "msg": "incoming request",
    "req": {
      "hostname": "localhost:80",
      "method": "GET",
      "remoteAddress": "127.0.0.1",
      "url": "/api/users/1",
    },
    "reqId": "req-1",
  },
  {
    "level": 30,
    "msg": "User with id 1 fetched",
    "reqId": "req-1",
  },
  {
    "level": 30,
    "msg": "request completed",
    "reqId": "req-1",
    "res": {
      "statusCode": 200,
    },
  },
]
`;

exports[`Users endpoint POST /users should create and return a user created: logs 1`] = `
[
  {
    "level": 30,
    "msg": "incoming request",
    "req": {
      "hostname": "localhost:80",
      "method": "POST",
      "remoteAddress": "127.0.0.1",
      "url": "/api/users",
    },
    "reqId": "req-1",
  },
  {
    "level": 30,
    "msg": "User with id",
    "reqId": "req-1",
    "user": {
      "email": "email@test.com",
      "name": "John Doe",
      "password": "password",
    },
  },
  {
    "level": 30,
    "msg": "request completed",
    "reqId": "req-1",
    "res": {
      "statusCode": 201,
    },
  },
]
`;
